gcc 11.2 error: implicit conversion from ‘ucs_status_t’ to ‘ucc_status_t’

UCC fails to compile on Ubuntu 21.10 with gcc 11.2.0, as UCC sets `-Werror` and `-Wenum-conversion` in the CFLAGS:
```console
make[3]: Entering directory '/home/fabecassis/github/ucc/src/components/tl/sharp'
  CC       libucc_tl_sharp_la-tl_sharp_context.lo
tl_sharp_context.c: In function ‘ucc_tl_sharp_context_t_init’:
tl_sharp_context.c:285:15: error: implicit conversion from ‘ucs_status_t’ to ‘ucc_status_t’ [-Werror=enum-conversion]
  285 |        status = ucc_rcache_create(&rcache_params, "SHARP", NULL, &self->rcache);
      |               ^
cc1: all warnings being treated as errors
make[3]: *** [Makefile:582: libucc_tl_sharp_la-tl_sharp_context.lo] Error 1
  CC       libucc_tl_sharp_la-tl_sharp_coll.lo
tl_sharp_coll.c: In function ‘ucc_tl_sharp_mem_register’:
tl_sharp_coll.c:87:16: error: implicit conversion from ‘ucs_status_t’ to ‘ucc_status_t’ [-Werror=enum-conversion]
   87 |         status = ucc_rcache_get(ctx->rcache, (void *)addr, length,
      |                ^
cc1: all warnings being treated as errors
make[3]: *** [Makefile:596: libucc_tl_sharp_la-tl_sharp_coll.lo] Error 1
```

UCX comes from HPC-X 2.9.0:
```console
$ ucx_info -v
# UCT version=1.11.0 revision 6031c98
# configured with: --disable-logging --disable-debug --disable-assertions --disable-params-check --with-knem --with-xpmem=/hpc/local/oss/xpmem/90a95a4 --without-java --enable-devel-headers --with-cuda=/hpc/local/oss/cuda11.2 --with-gdrcopy --prefix=/build-result/hpcx-v2.9.0-gcc-MLNX_OFED_LINUX-5.4-1.0.3.0-ubuntu20.04-x86_64/ucx
```

The workaround is simple for now though: `make CFLAGS=-Wno-error=enum-conversion`

--- a/src/components/tl/sharp/tl_sharp_coll.c
+++ b/src/components/tl/sharp/tl_sharp_coll.c
@@ -84,8 +84,7 @@ ucc_tl_sharp_mem_register(ucc_tl_sharp_context_t *ctx, void *addr,
     }
 
     if (ctx->rcache) {
-        status = ucc_rcache_get(ctx->rcache, (void *)addr, length,
-                                PROT_READ|PROT_WRITE, NULL,  &rregion);
+        status = ucc_rcache_get(ctx->rcache, (void *)addr, length, &rregion);
         if (status != UCC_OK) {
             tl_error(ctx->super.super.lib, "ucc_rcache_get failed");
             return UCC_ERR_INVALID_PARAM;

--- a/src/components/tl/sharp/tl_sharp_context.c
+++ b/src/components/tl/sharp/tl_sharp_context.c
@@ -282,7 +282,7 @@ UCC_CLASS_INIT_FUNC(ucc_tl_sharp_context_t,
        rcache_params.ops                = &ucc_tl_sharp_rcache_ops;
        rcache_params.flags              = 0;
 
-       status = ucc_rcache_create(&rcache_params, "SHARP", NULL, &self->rcache);
+       status = ucc_rcache_create(&rcache_params, "SHARP", &self->rcache);
        if (status != UCC_OK) {
            tl_error(self->super.super.lib, "failed to create rcache");
            status = UCC_ERR_NO_RESOURCE;

--- a/src/utils/ucc_rcache.h
+++ b/src/utils/ucc_rcache.h
@@ -15,11 +15,27 @@
 #define ucc_rcache_params_t          ucs_rcache_params_t
 #define ucc_rcache_region_t          ucs_rcache_region_t
 
-#define ucc_rcache_create            ucs_rcache_create
 #define ucc_rcache_destroy           ucs_rcache_destroy
-#define ucc_rcache_get               ucs_rcache_get
 #define ucc_rcache_region_hold       ucs_rcache_region_hold
 #define ucc_rcache_region_put        ucs_rcache_region_put
 #define ucc_rcache_region_invalidate ucs_rcache_region_invalidate
+
+/* Wrapper functions for status conversion */
+static inline ucc_status_t
+ucc_rcache_create(const ucc_rcache_params_t *params,
+                  const char *name, ucc_rcache_t **rcache_p)
+{
+    return ucs_status_to_ucc_status(ucs_rcache_create(
+                                        params, name, NULL, rcache_p));
+}
+
+static inline ucc_status_t
+ucc_rcache_get(ucc_rcache_t *rcache, void *address, size_t length,
+               ucc_rcache_region_t **region_p)
+{
+    return ucs_status_to_ucc_status(ucs_rcache_get(
+                                       rcache, address, length,
+                                       PROT_READ | PROT_WRITE, NULL, region_p));
+}
 
 #endif

